### PR TITLE
extsvc: Reuse rate limiters between clients for the same host

### PR DESCRIPTION
**Pull request by ryanslade**
This avoids creating a new source on every sync and more importantly makes
sure that we reuse the same rate limiter across syncs.

Part of https://github.com/sourcegraph/sourcegraph/issues/8546

### CODE REVIEW

**Comment by tsenart** [comment]
Yeah, we should find another way to share the rate limiter state, due to the reasons @mrnugget pointed out.

**Comment by ryanslade** [comment]
I know it's a hack but I'm going to focus on a proper rate limit solution next iteration

File: internal/extsvc/bitbucketcloud/client.go [hunk 1]
```diff
@@ -37,6 +38,13 @@ const (
 	RateLimitMaxBurstRequests  = 500
 )
 
+// Global limiter cache so that we reuse the same rate limiter for
+// the same code host, even between config changes.
+// The longer term plan is to have a rate limiter that is shared across
+// all services so the below is just a short term solution.
+var limiterMu sync.Mutex
+var limiterCache = make(map[string]*rate.Limiter)
```
Comment (unknwon) [inline internal/extsvc/bitbucketcloud/client.go]: Would use `sync.Map` is a cleaner solution? (without managing mutex ourselves). Just a suggestion.

**Comment by unknwon** [comment]
> I don't think this works without incoporating a cache buster whenever the configuration for an external service is updated.

Hmm, mind explain how could that happen? I saw only constants are used for instantiate a new limiter, and if API URL changes, the old limiter won't be used.

**Comment by ryanslade** [comment]
> > I don't think this works without incoporating a cache buster whenever the configuration for an external service is updated.
> 
> Hmm, mind explain how could that happen? I saw only constants are used for instantiating a new limiter, and if API URL changes, the old limiter won't be used.

We were caching the entire source, not just the rate limiter. So a change in config to the **source** wouldn't be picked up

File: internal/extsvc/bitbucketcloud/client.go (hunk 1 above)
Comment (keegancsmith) [reply internal/extsvc/bitbucketcloud/client.go]: To quote sync.Map docs:

> The Map type is specialized. Most code should use a plain Go map instead, with separate locking or coordination, for better type safety and to make it easier to maintain other invariants along with the map content.
>
> The Map type is optimized for two common use cases: (1) when the entry for a given key is only ever written once but read many times, as in caches that only grow, or (2) when multiple goroutines read, write, and overwrite entries for disjoint sets of keys. In these two cases, use of a Map may significantly reduce lock contention compared to a Go map paired with a separate Mutex or RWMutex.

This does match the first use case, but this map will be incredibly small for our case. I think the benefit of type safety + holding lock when there is a cache miss to create suggests we should stick with a map and mutex.

**Comment by ryanslade** [comment]
> Okay. But I'm wondering: why only BBS and Bitbucket Cloud?

That's the only place we currently use custom rate limiters

File: internal/extsvc/bitbucketcloud/client.go (hunk 1 above)
Comment (unknwon) [reply internal/extsvc/bitbucketcloud/client.go]: Thanks for referencing the docs, I guess it's more about personal taste in this case.

**Comment by unknwon** [comment]
> > > I don't think this works without incoporating a cache buster whenever the configuration for an external service is updated.
> > 
> > 
> > Hmm, mind explain how could that happen? I saw only constants are used for instantiating a new limiter, and if API URL changes, the old limiter won't be used.
> 
> We were caching the entire source, not just the rate limiter. So a change in config to the **source** wouldn't be picked up

Ok... that sounds we would have the same problem regardless of this PR?

**Comment by ryanslade** [comment]
The old code created a new `Source` for every sync so there was no chance of getting stale config. Once we added caching the risk was introduced

**Comment by unknwon** [comment]
> The old code created a new `Source` for every sync so there was no chance of getting stale config. Once we added caching the risk was introduced

Yes, but we use constants for instantiating rate limiters, no parameters from config, not the case?

**Comment by ryanslade** [comment]
The concern wasn't about the rate limit config, but about changes to the external service config. The change to cache the sources meant that we could cache stale config. Instead, we no only cache the rate limiter so that it's not reset on each sync.

**Comment by unknwon** [comment]
> The concern wasn't about the rate limit config, but about changes to the external service config. The change to cache the sources meant that we could cache stale config. Instead, we no only cache the rate limiter so that it's not reset on each sync.

Thanks for continuing answering my questions... and sorry I feel I miss some context, maybe next time in watercooler 😛

**Pull request merged**